### PR TITLE
Update TS rules

### LINKED ISSUE #12: Inline TypeScript config extends
https://github.com/MetaMask/eslint-config/blob/01d0bc9a7c5c449e7fe55474fe1859a89594d31f/config/typescript.js#L7-L11

The point of this config is to be what others extend from and to explicit about what's set (i.e. #1). These extends should be inlined as is possible.

### PR DESCRIPTION
**Pull request by whymarrh**
Closes #12

This PR explicitly sets all of the TS rules, removing the 3 `extends` from the config.

### CODE REVIEW

File: config/typescript.js [hunk 1]
```diff
@@ -1,20 +1,79 @@
...
     '@typescript-eslint/no-explicit-any': 'off',
+    '@typescript-eslint/no-extra-non-null-assertion': 'error',
+    '@typescript-eslint/no-extraneous-class': 'off',
+    '@typescript-eslint/no-inferrable-types': 'error',
+    '@typescript-eslint/no-misused-new': 'error',
+    '@typescript-eslint/no-namespace': ['error', { 'allowDefinitionFiles': true }],
+    '@typescript-eslint/no-non-null-asserted-optional-chain': 'error',
+    '@typescript-eslint/no-non-null-assertion': 'error',
+    '@typescript-eslint/no-parameter-properties': 'error',
+    '@typescript-eslint/no-require-imports': 'error',
```
Comment (rekmarks) [inline config/typescript.js]: What do we do in cases of untyped JavaScript imports?

For instance, there are a lot of cases in `gaba` and elsewhere where `require` is used to import both `json-rpc-engine` and other libraries, made by us or others.

File: config/typescript.js [hunk 2]
```diff
@@ -1,20 +1,79 @@
...
+    // Checked by TypeScript - ts(2376)
+    'no-this-before-super': 'off',
+    // Checked by TypeScript - `strictNullChecks`
+    'no-undef': 'off',
+    // Checked by TypeScript
+    'no-dupe-class-members': 'off',
+    // Checked by TypeScript
+    'no-redeclare': 'off',
+    // Checked by TypeScript
+    'import/no-unresolved': 'off',
```
Comment (rekmarks) [inline config/typescript.js]: This is to say, the TS compiler takes care of these?

Comment (whymarrh) [reply config/typescript.js]: Yup, the compiler should check all of these

**Comment by whymarrh** [comment]
> I realize these rules were already in effect in the config, but I have some questions about it and how to make existing code compatible.

Some of them were, I did add more rules. Sorry for being lazy and not clarifying that—I both inlined the existing extended configs and added the remainder of the non-type-info rules (see #2).

File: config/typescript.js (hunk 1 above)
Comment (rekmarks) [reply config/typescript.js]: If it's complicated, we can take that discussion offline. Worst case, I suppose we can disable this rule in the affected repos.

Comment (whymarrh) [reply config/typescript.js]: I think this is the correct _default_ but yes, this doesn't work for a lot of code. We have two options, really:

1. Easy option is to add a ignore comment:

```ts
// eslint-disable-line @typescript-eslint/no-require-imports
const EthjsQuery = require('ethjs-query');
```

It's ugly but so it should be. 🤷‍♂

2. Import it the "correct way" and add a bare-bones type definition file.

```ts
import EthjsQuery from 'ethjs-query';
```

```ts
// project-root/@types/ethjs-query.d.ts or something, I'm writing this from memory
declare module 'ethjs-query';
```

Comment (whymarrh) [reply config/typescript.js]: > both `json-rpc-engine` and other libraries, made by us or others.

The idea here is that this lint would get introduced as projects are "migrated", and where the migration is happening in dep tree order, libraries made by us should ship with their own type defns. The type defns that we write for libraries _made by others_ that aren't already on DefinitelyTyped can be shared across projects (more on that to come) but at worst each project could write their own barebones defns.

Comment (whymarrh) [reply config/typescript.js]: > Worst case, I suppose we can disable this rule in the affected repos as a stopgap.

That too, this is but a default config.

File: config/typescript.js [hunk 3]
```diff
@@ -1,20 +1,79 @@
...
   plugins: [
     '@typescript-eslint',
   ],
-  extends: [
-    'eslint:recommended',
-    'plugin:@typescript-eslint/eslint-recommended',
-    'plugin:@typescript-eslint/recommended',
-  ],
   rules: {
+    // Checked by TypeScript - ts(2378)
```
Comment (Gudahtt) [inline config/typescript.js]: Is there somewhere these redundant rules are listed?

Comment (whymarrh) [reply config/typescript.js]: I copied these from the plugin config itself, I don't believe there is a canonical list

File: config/typescript.js [hunk 4]
```diff
@@ -1,20 +1,79 @@
...
+    // Checked by TypeScript
+    'no-redeclare': 'off',
+    // Checked by TypeScript
+    'import/no-unresolved': 'off',
+
     // Rules from @typescript-eslint/eslint-plugin
+    '@typescript-eslint/adjacent-overload-signatures': 'error',
     '@typescript-eslint/array-type': 'error',
-    '@typescript-eslint/consistent-type-definitions': ['error', 'type'],
+    '@typescript-eslint/ban-ts-ignore': 'error',
```
Comment (Gudahtt) [inline config/typescript.js]: This rule is deprecated; it was replaced by `ban-ts-comment`

File: config/typescript.js [hunk 5]
```diff
@@ -1,20 +1,79 @@
...
+    '@typescript-eslint/ban-types': 'error',
+    '@typescript-eslint/class-name-casing': 'error',
```
Comment (Gudahtt) [inline config/typescript.js]: This rule is deprecated; it was replaced by `naming-convention`

File: config/typescript.js [hunk 6]
```diff
@@ -1,20 +1,79 @@
...
+    '@typescript-eslint/consistent-type-assertions': 'error',
+    '@typescript-eslint/consistent-type-definitions': ['error', 'interface'],
```
Comment (Gudahtt) [inline config/typescript.js]: Nit: `interface` is the default option

File: config/typescript.js [hunk 7]
```diff
@@ -1,20 +1,79 @@
...
+    '@typescript-eslint/explicit-function-return-type': 'off',
```
Comment (Gudahtt) [inline config/typescript.js]: This rule seems to be recommended. Any reason it was left off?

Comment (whymarrh) [reply config/typescript.js]: It's incredibly noisy and type inference + `noImplictAny` is a decent enough combo for now

**Comment by whymarrh** [comment]
There are rules that were intentionally omitted (as per my comment to Erik above, and #2)—any rules that require type information (as `await thenable` does) has been omitted.

**Comment by Gudahtt** [comment]
Ah; I thought the intention was to _add_ them all, but leave those rules disabled

**Comment by whymarrh** [comment]
Ah no, sorry, I wasn't planning on adding them in this PR, this was meant to resolve #12, mostly.

**Comment by whymarrh** [comment]
On deprecated rules, where is this list you're pulling from? The docs for the rules don't note that they're deprecated though some of them do seem to have a `deprecated` property (e.g. [this one](https://github.com/typescript-eslint/typescript-eslint/blob/9eab26f32221a314700eacafbc1c1c049cae110e/packages/eslint-plugin/src/rules/generic-type-naming.ts#L16)).

**Comment by Gudahtt** [comment]
Any that are omitted from the readme are deprecated; the list in the readme is automatically generated, and skips any deprecated rules. Unfortunately I couldn't find any list of deprecated rules.
https://github.com/typescript-eslint/typescript-eslint/tree/v2.20.0/packages/eslint-plugin

**Comment by whymarrh** [comment]
Aight I'll update these rules in another PR

**Comment by Gudahtt** [comment]
Though regarding #12, `await thenable` was recommended, so it does seem to fall under the scope of that issue. It was disabled in this PR.

File: config/typescript.js [hunk 8]
```diff
@@ -25,6 +84,9 @@ module.exports = {
     '@typescript-eslint/brace-style': 'error',
     'brace-style': 'off',
 
+    '@typescript-eslint/camelcase': 'error',
```
[the remark this answers is not in the document]
Comment (whymarrh) [reply config/typescript.js]: See #27

File: config/typescript.js (hunk 4 above)
Comment (whymarrh) [reply config/typescript.js]: See #27

File: config/typescript.js (hunk 5 above)
Comment (whymarrh) [reply config/typescript.js]: See #27

File: config/typescript.js (hunk 6 above)
Comment (whymarrh) [reply config/typescript.js]: See #27

**Comment by whymarrh** [comment]
See #27 where I've updated these rules to remove the unlisted ones and explicitly set the rules requiring type information to `'off'`.

**Pull request merged**